### ISSUE TITLE
template task fails when dest is a directory

### ISSUE THREAD

**Issue by nickswebsite** [ansible/ansible#10300]
The following playbook will fail:

```
- hosts: all
  sudo: yes
  tasks:
  - template: src=file.txt dest=/opt
```

The error code will be:

```
msg: failed to checksum remote file. Checksum error code: 3
```

/opt/file.txt does not exist after running the playbook.

This worked in ansible 1.8.2, so the issue was probably introduced within the last couple of days.

The 'copy' module doesn't have this problem.

**Comment by bcoca** [comment]
confirmed the issue, a quick workaround is adding a trailing slash to the directory, this works:

```
- hosts: all
  sudo: yes
  tasks:
      - template: src=file.txt dest=/opt/
```

**Comment by nickswebsite** [comment]
Yeah, I caught that.  Problem is, I'd have 100+ playbooks that would have to be updated with the workaround.  I doubt I'm the only one.

If you plan on fixing it in the next release, I'd rather not have to do that.  But, if this is the intended functionality going forward, I guess I'll have to.

Thanks for looking at it so soon!

**Comment by bcoca** [comment]
it is a bug, it will be fixed, I cannot promise next release though

**Comment by sysadmin75** [comment]
lib/ansible/runner/action_plugins/template.py
line 81:
        if dest.endswith("/"): # CCTODO: Fix path for Windows hosts.
            base = os.path.basename(source)
            dest = os.path.join(dest, base)

The easy fix (to me) would be to remove the if line and just leave (proper indention of course)
        base = os.path.basename(source)
        dest = os.path.join(dest, base)

But I'm unsure why the if statement was added. git blame shows:
21ba529f (Chris Church           2014-06-18 10:01:11 -0500  81)         if dest.endswith("/"): # CCTODO: Fix path for Windows hosts.

The commit statement:
Fixes/notes related to slashes in remote paths.

I don't see any reference to a bugfix/issue fix.

Anyone have any ideas why this was added?

**Comment by argami** [comment]
It also fails when using a file name

``` yml
- name: ufw before rules
  template: src=before.rules.j2 dest=/etc/ufw/before.rules owner=root group=root mode=0600
```

``` bash
failed: [server1] => {"failed": true}
msg: failed to checksum remote file. Checksum error code: 2

FATAL: all hosts have already failed -- aborting
```

**Comment by abadger** [comment]
Couple thoughts:

The commit that introduced the logic is actually here: c5d2f6b0d32b958bcd0365dbc341bb98a3aa3257  Only the comment was added in the later commit.  The log message on this commit doesn't explain things to me any better to me:

  implement lookup plugins for arbitrary enumeration over arbitrary things.   See the mailing list for some cool examples.

but I do note that there's similar changes in the copy module at the same time.

Second note:  I am guessing that this commit is what caused this problem to start occurring is here:

https://github.com/ansible/ansible-modules-core/commit/9e0959d600e5414b53eeb32ee1eba1a46d43daf6

(template calls copy to transfer the file to the remote system). Will start testing.

**Comment by abadger** [comment]
Okay, reverting that change does not fix this.

**Comment by abadger** [comment]
I'm able to reproduce the problem but I'm unable to reproduce it working in 1.8.2.  Can someone confirm that these playbooks work there?

**Comment by abadger** [comment]
Okay --- an up update to my last message -- `- template: src=file.txt dest=/opt` does work in 1.8.2.

`template: src=before.rules.j2 dest=/etc/ufw/before.rules owner=root group=root mode=0600` does not.

So this looks like two separate problems despite the similarities.

**Comment by nickswebsite** [comment]
I can confirm that the one listed in my original report works in 1.8.2.  I have several playbooks that work using template and copy.

I was not able to reproduce it with copy.  I went poking around the source code looking for the module, but files/template.py was 'entirely implemented server side' and I wasn't really sure where to find the actual implementation.

**Comment by abadger** [comment]
git bisect shows that this is the commit that broke template for the original report after 1.8.2 https://github.com/ansible/ansible/commit/2a288141d3d3ea709d83772b431f6d58dae22198

**Comment by abadger** [comment]
@nickswebsite -- https://github.com/ansible/ansible/commit/ac40ffa157f39d7fdd5dba70ed8f5ae5df7eaa62 should fix the issue you reported.  I'm boarding a plane right now but I'll be sure to check that our integration tests passed in our automated system once I land.

@argaml this change does not fix your error.  I'll take a look once I land and let you know if you should open a separate ticket.

# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!

**Comment by abadger** [comment]
@argami Okay, I diagnosed that my test of your issue was failing because I did not have an /etc/ufw directory.  After creating that, I was unable to reproduce your issue.  I do note that in your comment it mentions  `Checksum error code: 2` Error code 2 happens when there's a permission issue.  Perhaps  /etc/ufw is writable only by root on your remote box but you aren't using sudo to switch from the user account you ssh'd into the box as to the root account?

If you continue to have issues, either post on the mailing list or open a new ticket and we can diagnose further.
